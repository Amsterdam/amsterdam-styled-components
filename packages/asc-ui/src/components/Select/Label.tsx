import React from 'react'
import { AscCore } from '@datapunt/asc-core'
import { ReactComponent as Elipsis } from '@datapunt/asc-assets/lib/Icons/Elipsis.svg'
import { ReactComponent as ChevronTop } from '@datapunt/asc-assets/lib/Icons/Chevron-Top.svg'
import { ReactComponent as ChevronDown } from '@datapunt/asc-assets/lib/Icons/Chevron-Down.svg'
import { Icon } from '../..'

const Label = ({ id, label, open, onOpenClose }: any) => {
  return (
    <AscCore.Select.Label.default
      {...{ id }}
      onBlur={() => onOpenClose(false)}
      onClick={() => onOpenClose()}
    >
      {label || (
        <Icon inline>
          <Elipsis />
        </Icon>
      )}
      <Icon inline>{open ? <ChevronDown /> : <ChevronTop />}</Icon>
    </AscCore.Select.Label.default>
  )
}

export default Label
