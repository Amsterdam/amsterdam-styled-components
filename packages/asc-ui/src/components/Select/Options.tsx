import React from 'react'
import { AscCore } from '@datapunt/asc-core'

type Props = {
  children: React.ReactChildren & {
    props: AscCore.Select.Option.Props
  }
  currentTab: string
}

const Options = ({ children: childrenProp, currentTab }: Props) => {
  let childIndex = 0
  const children = React.Children.map(childrenProp, child => {
    if (!React.isValidElement(child)) {
      return null
    }

    const { label } = child.props
    const childValue = label === undefined ? childIndex : label
    const selected = childValue === currentTab

    childIndex += 1
    return React.cloneElement(child as React.ReactElement<any>, {
      isCurrentTab: selected,
    })
  })
  return <>{children}</>
}

export default Options
