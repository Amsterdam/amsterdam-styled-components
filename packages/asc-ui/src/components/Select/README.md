# Select

[Design System](https://designsystem.amsterdam.nl/7awj1hc9f/p/6561a5-selection-dropdown)

**Component status: experimental**
