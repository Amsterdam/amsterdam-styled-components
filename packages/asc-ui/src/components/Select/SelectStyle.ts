import { ChevronDown } from '@amsterdam/asc-assets'
import { CSSProperties } from 'react'
import styled, { css } from 'styled-components'
import { themeColor, themeSpacing } from '../../utils'
import { INPUT_SIZE } from '../shared/constants'

export type Props = {
  id?: string
  value?: string
  label?: string
  disabled?: boolean
  srOnly?: boolean
  error?: boolean | string
  errorMessage?: string
  labelStyle?: CSSProperties
  errorStyle?: CSSProperties
}

export const SelectWrapper = styled.div`
  position: relative;
  height: ${INPUT_SIZE}px;
  width: 100%;
`

/**
 * Contains all of the absolute content which is placed above the select element.
 */
export const AbsoluteContentWrapper = styled.div.attrs({
  'aria-hidden': 'true',
})`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 3px; /* Allows the default browser(s) default focus style to be displayed. */
  padding: 0 calc(${themeSpacing(3)} - 1px); /* Match the spacing of the select element. */
  display: flex;
  align-items: center;
  pointer-events: none;
  background-color: ${themeColor('tint', 'level1')};
`

/**
 * Since Chrome does not support `text-overflow: hidden` we need show the value in a separate element.
 */
export const SelectedValue = styled.div<{ disabled?: boolean }>`
  margin-right: auto;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
    `}
`
export const SelectIcon = styled(ChevronDown)`
  display: block;
  width: ${themeSpacing(3)};
  height: ${themeSpacing(3)};
  margin-left: ${themeSpacing(3)};
`

const SelectStyle = styled.select<Props>`
  width: 100%;
  height: 100%;
  padding: ${themeSpacing(2, 3)};
  font-size: 1rem;
  border: 1px solid ${themeColor('tint', 'level5')};
  border-radius: 0;
  background-color: ${themeColor('tint', 'level1')};
  appearance: none;
  cursor: pointer;

  /* IE11 (hide native arrow button) */
  &::-ms-expand {
    display: none;
  }

  /* IE11 (hide native focus state) */
  &::-ms-value {
    color: black;
    background: transparent;
  }

  /* Firefox (hide native focus state) */
  &:-moz-focusring {
    color: transparent;
    text-shadow: 0 0 0 ${themeColor('tint', 'level7')};
  }

  &:hover {
    ${({ disabled, error }) =>
      !disabled &&
      !error &&
      css`
        border: 1px solid ${themeColor('tint', 'level7')};
      `}
  }

  &:disabled {
    pointer-events: none;
    opacity: 0.5;
    border: 1px solid ${themeColor('tint', 'level4')};
  }

  ${({ error }) =>
    error &&
    css`
      border: 1px solid ${themeColor('error', 'main')};
    `}
`

export default SelectStyle
