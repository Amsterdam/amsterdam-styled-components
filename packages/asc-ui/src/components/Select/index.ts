export { default as SelectStyle } from './SelectStyle'
export { default } from './Select'
