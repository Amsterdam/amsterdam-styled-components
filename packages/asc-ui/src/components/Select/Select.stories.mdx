import { createRef, useEffect } from 'react'
import { Meta, Story, Preview } from '@storybook/addon-docs/blocks'
import Select from './Select'
import ErrorMessage from '../ErrorMessage'

export const SelectWithRef = () => {
  const ref = createRef()
  useEffect(() => {
    const element = ref.current
    function onChange(e) {
      const { value } = e.target
      // eslint-disable-next-line no-console
      console.log('Value changed!', value)
    }
    if (!element) {
      return undefined
    }
    element.addEventListener('change', onChange)
    return () => {
      element.removeEventListener('change', onChange)
    }
  }, [ref])
  return (
    <Select id="with-ref" label="Select with ref" ref={ref}>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  )
}

<Meta
  title="Experimental/Atoms/Select (Dropdown)"
  component={Select}
  decorators={[
    (storyFn) => (
      <div
        style={{
          width: '400px',
          padding: '40px 10px',
        }}
      >
        {storyFn()}
      </div>
    ),
  ]}
/>

# Select (dropdown)

Use this component as a last resort, because some users are having trouble
with it. Try to use Radio Buttons first, or in case of a long list, try
to ask multiple questions (text from Design System)

[Design System](https://designsystem.amsterdam.nl/7awj1hc9f/p/6561a5-selection-dropdown)

**Component status: experimental**

## Default state

<Preview>
  <Story name="Default state">
    <Select id="select-1" label="Label">
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </Story>
</Preview>

## Selected state

<Preview>
  <Story name="Selected state">
    <Select id="select-1" label="Label">
      <option value="1">Option 1</option>
      <option value="2" selected>
        Option 2
      </option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </Story>
</Preview>

## Disabled state

<Preview>
  <Story name="Disabled state">
    <React.Fragment>
      <Select id="select-1" label="Disabled Select component" disabled>
        <option value="1">Option 1</option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
      <br />
      <br />
      <Select id="select-2" label="Disabled specific Option">
        <option value="1" disabled>
          Option 1
        </option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
    </React.Fragment>
  </Story>
</Preview>

## Error state

<Preview>
  <Story name="Error state">
    <React.Fragment>
      <Select id="select-1" label="With Error Message" error>
        <option value="1">Option 1</option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
      <ErrorMessage message="Error message" />
      <br />
      <br />
      <Select id="select-2" label="Without Error Message" error>
        <option value="1">Option 1</option>
        <option value="2">Option 2</option>
        <option value="3">Option 3</option>
        <option value="4">Option 4</option>
      </Select>
    </React.Fragment>
  </Story>
</Preview>

## With ref

<Preview>
  <Story name="With ref">
    <SelectWithRef />
  </Story>
</Preview>
