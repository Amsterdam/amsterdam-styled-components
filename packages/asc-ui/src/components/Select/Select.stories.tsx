import React from 'react'
import Select from './Select'

export default {
  title: 'Experimental/Atoms/Select (Dropdown)',

  decorators: [
    (storyFn: () => React.ReactNode) => (
      <div
        style={{
          display: 'flex',
          maxWidth: 400,
          flexDirection: 'column',
          alignItems: 'flex-start',
          padding: '40px 10px',
        }}
      >
        {storyFn()}
      </div>
    ),
  ],
}

export const Default = () => (
  <>
    <Select id="select-1" label="With label">
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>

    <br />
    <br />
    <em>Without label</em>
    <Select id="select-2">
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </>
)

export const Disabled = () => (
  <>
    <Select id="select-1" label="Disabled" disabled>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
    <br />
    <br />
    <Select id="select-2" label="Specific disabled">
      <option value="1" disabled>
        Option 1
      </option>
      <option value="2" disabled>
        Option 2
      </option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </>
)

export const Error = () => (
  <>
    <Select id="select-1" label="Label" errorMessage="Error message">
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
    <br />
    <br />
    <Select id="select-2" label="Label only" error>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
    <br />
    <br />
    <Select id="select-3" errorMessage="Error message only">
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
    <br />
    <br />
    <em>No labels at all</em>
    <Select id="select-4" error>
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
      <option value="4">Option 4</option>
    </Select>
  </>
)
