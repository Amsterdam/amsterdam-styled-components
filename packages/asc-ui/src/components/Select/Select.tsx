import React from 'react'
import FormLabelStyle from '../FormLabelStyle'
import SelectStyle, { Props, SelectIcon, SelectWrapper } from './SelectStyle'

const Select = React.forwardRef<
  HTMLSelectElement,
  Props & React.HTMLAttributes<HTMLSelectElement>
>(
  (
    { id, value, label, srOnly: srOnlyProp, error, labelStyle, ...otherProps },
    ref,
  ) => {
    const srOnly = srOnlyProp || false

    return (
      <>
        {label && (
          <FormLabelStyle
            srOnly={srOnly}
            htmlFor={id}
            label={label}
            style={labelStyle}
          >
            {label}
          </FormLabelStyle>
        )}
        <SelectWrapper>
          <SelectStyle
            {...{
              ...otherProps,
              id,
              value,
              error,
            }}
            ref={ref}
          />
          <SelectIcon />
        </SelectWrapper>
      </>
    )
  },
)

export default Select
