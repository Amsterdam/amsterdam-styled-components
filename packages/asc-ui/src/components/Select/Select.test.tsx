/* eslint-disable @typescript-eslint/ban-types */
import { createRef } from 'react'
import { render, fireEvent, act } from '@testing-library/react'
import 'jest-styled-components'
import Select from './Select'
import Label from '../Label'
import { ThemeProvider } from '../../theme'

describe('Select', () => {
  let container: any
  let rerender: Function
  let unmount: Function
  let getByTestId: Function

  const onChangeMock = jest.fn()
  const onKeyDownMock = jest.fn()

  beforeEach(() => {
    onChangeMock.mockReset()
    onKeyDownMock.mockReset()
    ;({ container, rerender, unmount, getByTestId } = render(
      <ThemeProvider>
        <Label htmlFor="select-1" label="Select 1" id="label-1">
          <Select
            id="select-1"
            onChange={onChangeMock}
            onKeyDown={onKeyDownMock}
          >
            <option value="1">Option 1</option>
            <option value="2">Option 2</option>
            <option value="3">Option 3</option>
            <option value="4">Option 4</option>
          </Select>
        </Label>
      </ThemeProvider>,
    ))
  })

  it('should match the style', () => {
    const select = container.querySelector('#select-1')
    expect(select).toHaveStyleRule('width', '100%')
  })

  it('should trigger the change event', () => {
    const select = container.querySelector('#select-1')

    expect(select.value).toBe('1')

    select.value = '2'

    expect(onChangeMock).not.toHaveBeenCalled()

    act(() => {
      fireEvent.change(select)
    })

    expect(select.value).toBe('2')
    expect(onChangeMock).toHaveBeenCalledTimes(1)
  })

  it('should trigger the keyDown event', () => {
    const select = container.querySelector('#select-1')

    expect(onKeyDownMock).not.toHaveBeenCalled()

    act(() => {
      fireEvent.keyDown(select, { key: 'Enter', code: 13 })
    })

    expect(onKeyDownMock).toHaveBeenCalledTimes(1)
  })

  it('should trigger the onChange event', () => {
    const select = container.querySelector('#select-1')

    expect(onChangeMock).not.toHaveBeenCalled()

    const firstOption = select.options[0]

    act(() => {
      fireEvent.change(select, { target: { value: firstOption.value } })
    })

    expect(onChangeMock).toHaveBeenCalledTimes(1)
  })

  it("should update the overlay label when the field's value changes", () => {
    // clean the DOM by unmounting the component
    unmount()

    rerender(
      <ThemeProvider>
        <Label htmlFor="select-2" label="Select 2" id="label-2">
          <Select
            id="select-2"
            onChange={onChangeMock}
            onKeyDown={onKeyDownMock}
          >
            <option value="11">Option 11</option>
            <option value="12">Option 12</option>
            <option value="13">Option 13</option>
            <option value="14">Option 14</option>
          </Select>
        </Label>
      </ThemeProvider>,
    )

    const select = container.querySelector('#select-2')
    const firstOption = select.options[0]
    const secondOption = select.options[1]

    // first option value will always be set as the overlay label
    expect(getByTestId('selectedValue')).toHaveTextContent(firstOption.value)

    // programatically changing the value
    rerender(
      <ThemeProvider>
        <Label htmlFor="select-2" label="Select 2" id="label-2">
          <Select
            id="select-2"
            onChange={onChangeMock}
            onKeyDown={onKeyDownMock}
            value={secondOption.value}
          >
            <option value="11">Option 11</option>
            <option value="12">Option 12</option>
            <option value="13">Option 13</option>
            <option value="14">Option 14</option>
          </Select>
        </Label>
      </ThemeProvider>,
    )

    expect(getByTestId('selectedValue')).toHaveTextContent(secondOption.value)
  })

  it('should handle refs', () => {
    const ref = createRef<HTMLSelectElement>()

    render(
      <ThemeProvider>
        <Select ref={ref} />
      </ThemeProvider>,
    )

    expect(ref.current).toBeInstanceOf(HTMLSelectElement)
  })
})
