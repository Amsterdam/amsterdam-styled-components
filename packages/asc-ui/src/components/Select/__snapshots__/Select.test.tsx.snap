// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render 1`] = `
.c4 {
  position: absolute;
  display: block;
  width: 12px;
  height: 12px;
  top: calc(50% - 6px);
  right: 12px;
}

.c2 {
  position: relative;
  height: 40px;
  width: 100%;
}

.c3 {
  width: 100%;
  height: 100%;
  padding: 0 12px;
  font-size: 1rem;
  border: 0;
  border-radius: 0;
  background-color: #ffffff;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  box-shadow: inset 0 0 0 1px #767676;
}

.c3::-ms-expand {
  display: none;
}

.c3::-ms-value {
  color: black;
  background: transparent;
}

.c3:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000000;
}

.c3:hover {
  box-shadow: inset 0 0 0 1px #000000;
}

.c3:focus {
  outline-color: #FEC813;
  outline-style: solid;
  outline-offset: 0px;
  outline-width: 2px;
}

.c3:disabled {
  pointer-events: none;
  opacity: 0.5;
  box-shadow: inset 0 0 0 1px #B4B4B4;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  color: #000000;
  cursor: pointer;
}

.c0 .c5 {
  padding-right: 12px;
}

.c1 {
  margin: 9.2px 0;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

<div>
  <label
    class="c0"
    for="select-1"
    id="label-1"
  >
    <span
      class="c1"
    >
      Select 1
    </span>
    <div
      class="c2"
    >
      <select
        class="c3"
        id="select-1"
      >
        <option
          value="1"
        >
          Option 1
        </option>
        <option
          value="2"
        >
          Option 2
        </option>
        <option
          value="3"
        >
          Option 3
        </option>
        <option
          value="4"
        >
          Option 4
        </option>
      </select>
      <svg
        class="c4"
        focusable="false"
        viewBox="0 0 32 32"
      >
        <path
          d="M16 25.757l-16-16 2.91-2.9L16 19.937l13.09-13.08 2.91 2.9z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </label>
</div>
`;
